File: 39-combination-sum/combination-sum.cpp
class Solution {
public:
    void f(int idx, vector<int>& candidates, vector<int>& temp, int target,
           vector<vector<int>>& result) {
      
        if (idx  < 0) {
            if (target  == 0) {
                result.push_back(temp);  
            }
            return;
        }

        // pick

        if (target >= candidates[idx]) {

            temp.push_back(candidates[idx]);
            f(idx, candidates, temp, target - candidates[idx], result);

            temp.pop_back();
        }

        f(idx - 1, candidates, temp, target, result);
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> result;
        vector<int> temp;
        int n = candidates.size();
        f(n - 1, candidates, temp, target, result);
        return result;
    }
};

File: 39-combination-sum/combination-sum.java
class Solution {
    static void f(int idx,int[] candidates, List<Integer>temp,int target, List<List<Integer>>ans){
        int n=candidates.length;
        if(idx==n){
            if(target == 0){
                ans.add(new ArrayList<>(temp));
            }
            return;
        }

        //pick

        int val=candidates[idx];
        if(val <= target){
            temp.add(val);
            f(idx,candidates,temp,target-val,ans);
            temp.remove(temp.size()-1);
        }
        f(idx+1,candidates,temp,target,ans);

    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>>ans =new ArrayList<>();
        List<Integer>temp=new ArrayList<>();
        
        f(0,candidates,temp, target,ans);
        return ans;
        
        
    }
}